### DISCUSSION TITLE
Enforcing SSL connection to database

### DISCUSSION THREAD

**Discussion by PiotrWachulec** [YOURLS/YOURLS#3795] [Q&A]
Hello Everyone!

I'm playing with Yourls for my needs. I used the official Docker image to simplify deployment. When I tried to run this I got the error that I can't connect to the MySQL database because it is enforcing a SSL connection.

Is there any option in the configuration to enforce at the Yourls connection to db using SSL? For my test purposes, I turned off dropping non-SSL connections, but I would like to avoid such configuration on production.

Many thanks in advance!
Piotr

P. S. Sorry if it was discussed, I tried to check but I haven't found proper for me answer.

**Comment by logusivam** [comment]
YOURLS does not have a built-in `YOURLS_DB_SSL` constant or a config option for SSL database connections in its standard `user/config.php` — this is a genuine gap in the current configuration API. However, you can fully enable SSL for the MySQL connection through YOURLS's database bootstrap hooks without patching core, and it works correctly with the official Docker image.

---

## Why this happens

The official YOURLS Docker image uses the standard `user/config.php` → PDO connection path. YOURLS constructs the PDO DSN and passes connection options via `YOURLS_DB_*` constants, but there is no built-in constant for SSL PDO attributes (`PDO::MYSQL_ATTR_SSL_CA`, etc.). When your MySQL server enforces SSL and YOURLS connects without SSL attributes, the connection is rejected.

---

## Fix — Inject SSL PDO attributes via a must-use plugin

YOURLS fires a hook just before the database connection is established. You can use this to pass the SSL certificate attributes directly to PDO.

### Step 1 — Obtain your SSL certificate files

You need one or more of these from your MySQL server administrator or cloud provider:

| File | PDO attribute | Required? |
|------|--------------|----------|
| CA certificate (`ca-cert.pem`) | `PDO::MYSQL_ATTR_SSL_CA` | Yes (most setups) |
| Client certificate (`client-cert.pem`) | `PDO::MYSQL_ATTR_SSL_CERT` | Only for mutual TLS |
| Client key (`client-key.pem`) | `PDO::MYSQL_ATTR_SSL_KEY` | Only for mutual TLS |

For most managed MySQL services (AWS RDS, DigitalOcean, PlanetScale, etc.) you only need the CA certificate.

### Step 2 — Place certificates inside the Docker container

In your `docker-compose.yml`, mount a local certs folder into the container:

```yaml
services:
  yourls:
    image: yourls:latest
    volumes:
      - ./yourls-config/user:/var/www/html/user    # your user/ folder
      - ./mysql-certs:/etc/mysql-certs              # SSL certificates
    environment:
      YOURLS_DB_HOST: your-mysql-host
      YOURLS_DB_USER: yourls
      YOURLS_DB_PASS: yourpassword
      YOURLS_DB_NAME: yourls
      YOURLS_SITE:    https://yourdomain.com
      YOURLS_USER:    admin
      YOURLS_PASS:    adminpassword
```

### Step 3 — Create the SSL plugin

Create `user/plugins/db-ssl/plugin.php`:

```php
<?php
/*
Plugin Name: Database SSL Connection
Plugin URI: https://yourls.org/
Description: Forces YOURLS database connection to use SSL via PDO attributes
Version: 1.0
Author: You
*/

if ( !defined( 'YOURLS_ABSPATH' ) ) die();

yourls_add_filter( 'db_connect_options', 'dbssl_add_pdo_ssl_options' );
function dbssl_add_pdo_ssl_options( $options ) {

    // Path to your CA certificate inside the container
    $options[ PDO::MYSQL_ATTR_SSL_CA ]            = '/etc/mysql-certs/ca-cert.pem';

    // Uncomment below only if your MySQL requires mutual TLS (client cert + key)
    // $options[ PDO::MYSQL_ATTR_SSL_CERT ]        = '/etc/mysql-certs/client-cert.pem';
    // $options[ PDO::MYSQL_ATTR_SSL_KEY ]         = '/etc/mysql-certs/client-key.pem';

    // Set to false only for testing — ALWAYS keep true in production
    $options[ PDO::MYSQL_ATTR_SSL_VERIFY_SERVER_CERT ] = true;

    return $options;
}
```

Activate it from **Admin → Manage Plugins**.

---

## Alternative — Set SSL via DSN in `user/config.php` directly

If you prefer not to use a plugin, you can append SSL parameters directly to the DSN string in `user/config.php`. Add this after your standard `define` constants:

```php
// Standard YOURLS DB constants
define( 'YOURLS_DB_HOST', 'your-mysql-host' );
define( 'YOURLS_DB_USER', 'yourls' );
define( 'YOURLS_DB_PASS', 'yourpassword' );
define( 'YOURLS_DB_NAME', 'yourls' );

// SSL — override the PDO connection with SSL attributes
// This runs after YOURLS sets up its DB object
global $ydb;
add_action( 'init', function() {
    global $ydb;
    if ( $ydb && method_exists( $ydb, 'setAttribute' ) ) {
        $ydb->setAttribute( PDO::MYSQL_ATTR_SSL_CA,  '/etc/mysql-certs/ca-cert.pem' );
        $ydb->setAttribute( PDO::MYSQL_ATTR_SSL_VERIFY_SERVER_CERT, true );
    }
});
```

> ⚠️ The plugin approach (Step 3) is cleaner and more reliable because it hooks before the connection is established rather than after. The DSN approach risks a race condition where the connection attempt happens before the attributes are set.

---

## Verifying the SSL connection works

After activating the plugin, verify the connection is actually using SSL by running this query in phpMyAdmin or your MySQL client:

```sql
SHOW STATUS LIKE 'Ssl_cipher';
```

If the `Value` column shows a cipher string (e.g. `TLS_AES_256_GCM_SHA384`), the connection is encrypted. If it is empty, SSL is not active.

---

## Production checklist

| Item | Recommendation |
|------|---------------|
| `PDO::MYSQL_ATTR_SSL_VERIFY_SERVER_CERT` | Always `true` in production — `false` defeats the purpose of SSL |
| Certificate expiry | Set a calendar reminder to rotate before expiry |
| Cert file permissions | `chmod 600` on all `.pem` files inside the container |
| Docker secrets | Consider using Docker secrets instead of bind-mounting cert files for production deployments |

---

## On the feature gap

You're right that this should be a first-class YOURLS config option. A `YOURLS_DB_SSL_CA` constant (and equivalents for cert/key) would make this accessible to non-developer users without needing a plugin. It's worth opening a feature request in the Ideas category — the implementation would be a small addition to the DB bootstrap code and would benefit anyone using managed MySQL services that enforce SSL.

Hope this helps — if it answers your question, would you mind clicking **"Mark as answer"** so others searching for the same thing can find it quickly?
